transport/tlsfrag/writer.go

```diff
@@ -0,0 +1,177 @@
+		if _, err = w.flushRecord(); err != nil {
+			return
+		}
+	}
+
+	if len(p) > 0 {
+		m, e := w.base.Write(p)
+		n += m
+		err = e
+	}
+	return
+}
+
+// ReadFrom implements io.ReaderFrom.ReadFrom. It attempts to split the first packet into two TLS records if the data
+// corresponds to a TLS Client Hello record. And then copies the remaining data from r to the base io.Writer until EOF
+// or error.
+//
+// If the first packet is not a valid TLS Client Hello, everything from r gets copied to the base io.Writer as is.
+//
+// It returns the number of bytes read. Any error except EOF encountered during the read is also returned.
+func (w *clientHelloFragReaderFrom) ReadFrom(r io.Reader) (n int64, err error) {
+	if !w.done {
+		// not yet splitted, append to the buffer
+		if w.record == nil {
+			if n, err = w.helloBuf.ReadFrom(r); err == nil {
+				// EOF, but Client Hello is not fully received yet
+				return
+			}
+			if errors.Is(err, errTLSClientHelloFullyReceived) {
+				w.splitHelloBufToRecord()
+			} else {
+				w.copyHelloBufToRecord()
+			}
+		}
+		// already splitted (but previous Writes might fail), try to flush all remaining w.record to w.base
+		if _, err = w.flushRecord(); err != nil {
+			return
+		}
+	}
+
+	m, e := w.baseRF.ReadFrom(r)
+	n += m
+	err = e
+	return
+}
+
+// copyHelloBufToRecord copies w.helloBuf into w.record without allocations.
+func (w *clientHelloFragWriter) copyHelloBufToRecord() {
+	w.record = bytes.NewBuffer(w.helloBuf.Bytes())
+	w.helloBuf = nil // allows the GC to recycle the memory
+}
+
+// splitHelloBufToRecord splits w.helloBuf into two records and put them into w.record without allocations.
+func (w *clientHelloFragWriter) splitHelloBufToRecord() {
+	received := w.helloBuf.Bytes()
+	content := received[recordHeaderLen:]
+	split := w.frag(content)
+	if split <= 0 || split >= len(content) {
+		w.copyHelloBufToRecord()
+		return
+	}
+
+	// received: | <== header (5) ==> | <== split ==> | <== len(content)-split ==> |  ... cap with padding (5) ... |
+	//                                                 \                            \
+	//                                                  +-----------------+          +-----------------+
+	//                                                                     \                            \
+	// splitted: | <== header (5) ==> | <== split ==> | <== header2 (5) ==> | <== len(content)-split ==> |
+	splitted := received[:len(received)+recordHeaderLen]
+	hdr1 := tlsRecordHeaderFromRawBytes(splitted[:recordHeaderLen])
+	hdr2 := tlsRecordHeaderFromRawBytes(splitted[recordHeaderLen+split : recordHeaderLen*2+split])
+	recvContent2 := splitted[recordHeaderLen+split : len(received)]
+	content2 := splitted[recordHeaderLen*2+split:]
+	copy(content2, recvContent2)
+	copy(hdr2, hdr1)
+	hdr1.SetPayloadLen(uint16(split))
+	hdr2.SetPayloadLen(uint16(len(content) - split))
+	w.record = bytes.NewBuffer(splitted)
```

Comment:
Why don't you write to the base writer here?

Reply:
A single `Write` might fail, so you still need to track the remaining bytes somewhere in `b`, and you need to `Write` them again in the following `Write`s or `ReadFrom`s issued by the caller.

I think it makes more sense to separate the "flush bytes to base" logic from the `split`.